src/interface/guide/foundation/analyzers/MeleeUptimeAnalyzer.tsx

```diff
@@ -0,0 +1,159 @@
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, { AnyEvent, CastEvent, EndChannelEvent, FightEndEvent } from 'parser/core/Events';
+import SPELLS from 'common/SPELLS';
+import Haste from 'parser/shared/modules/Haste';
+import Spell from 'common/SPELLS/Spell';
+import { QualitativePerformance } from 'parser/ui/QualitativePerformance';
+import ROLES from 'game/ROLES';
+import SPECS, { Spec } from 'game/SPECS';
+
+interface MeleeCast {
+  event: CastEvent;
+  haste: number;
+}
+
+const MAX_MELEE_TIMER = 4000;
+const MIN_MELEE_TIMER = 800;
+const MELEE_HEALERS = [SPECS.MISTWEAVER_MONK.id, SPECS.HOLY_PALADIN.id];
+
+export class MeleeUptimeAnalyzer extends Analyzer.withDependencies({ haste: Haste }) {
+  private recentMelees: MeleeCast[] = [];
+  private expectedNextMeleeTimestamp: number | undefined = undefined;
+
+  private gaps: Array<{ start: number; end: number }> = [];
+
+  public static withMeleeAbility(ability: Spell) {
+    return class extends MeleeUptimeAnalyzer {
+      constructor(options: Options) {
+        super(options, ability);
+      }
+    };
+  }
+
+  static isMeleeSpec(spec?: Spec): boolean {
+    if (!spec) {
+      return false;
+    }
+
+    return (
+      spec.role === ROLES.DPS.MELEE || spec.role === ROLES.TANK || MELEE_HEALERS.includes(spec.id)
+    );
+  }
+
+  constructor(options: Options, meleeAbility?: Spell) {
+    super(options);
+
+    if (!MeleeUptimeAnalyzer.isMeleeSpec(options.owner.config.spec)) {
+      this.active = false;
+      return;
+    }
+
+    this.addEventListener(
+      Events.cast
+        .by(SELECTED_PLAYER)
+        .spell(meleeAbility ? [SPELLS.MELEE, meleeAbility] : SPELLS.MELEE),
+      this.onMeleeCast,
+    );
+    this.addEventListener(Events.EndChannel.by(SELECTED_PLAYER), this.onChannelEnd);
+    this.addEventListener(Events.fightend, this.onFightEnd);
+  }
+
+  public get meleeUptimeGaps(): typeof this.gaps {
+    return this.gaps;
+  }
+
+  public get meleeUptimePercentage(): number {
+    const downtime = this.gaps.reduce((total, gap) => total + gap.end - gap.start, 0);
+    return (this.owner.fightDuration - downtime) / this.owner.fightDuration;
+  }
+
+  public get meleeUptimePerformance(): QualitativePerformance {
```

Comment:
Consider using `evaluateQualitativePerformanceByThreshold` instead of manually doing calcs